viscy/representation/evaluation/distance.py

```diff
@@ -20,67 +21,184 @@ def calculate_cosine_similarity_cell(embedding_dataset, fov_name, track_id):
-        }
-        return mean_displacement_per_tau, std_displacement_per_tau
-
-    return displacement_per_tau
+    # Process each track
+    for fov_name, track_id in unique_tracks:
+        # Get sorted track data
+        mask = (fov_names == fov_name) & (track_ids == track_id)
+        times = timepoints[mask]
+        track_embeddings = embeddings[mask]
+
+        # Sort by time
+        time_order = np.argsort(times)
+        times = times[time_order]
+        track_embeddings = track_embeddings[time_order]
+
+        # Process each time point
+        for t_idx, t in enumerate(times[:-1]):
+            current_embedding = track_embeddings[t_idx]
+
+            # Check all possible future time points
+            for future_idx, future_time in enumerate(
+                times[t_idx + 1 :], start=t_idx + 1
+            ):
+                tau = future_time - t
+                future_embedding = track_embeddings[future_idx]
+
+                if distance_metric in ["cosine", "cosine_dissimilarity"]:
+                    dot_product = np.dot(current_embedding, future_embedding)
+                    norms = np.linalg.norm(current_embedding) * np.linalg.norm(
+                        future_embedding
+                    )
+                    similarity = dot_product / norms
+                    displacement = (
+                        1 - similarity
+                        if distance_metric == "cosine_dissimilarity"
+                        else similarity
+                    )
+                else:  # Euclidean metrics
+                    diff_squared = np.sum((current_embedding - future_embedding) ** 2)
+                    displacement = (
+                        diff_squared
+                        if distance_metric == "euclidean_squared"
+                        else np.sqrt(diff_squared)
+                    )
+
+                displacement_per_tau[int(tau)].append(displacement)
+
+    return dict(displacement_per_tau)
+
+
+def compute_displacement_statistics(
+    displacement_per_tau: Dict[int, List[float]]
+) -> Tuple[Dict[int, float], Dict[int, float]]:
+    """Compute mean and standard deviation of displacements for each tau.
+
+    Parameters
+    ----------
+    displacement_per_tau : Dict[int, List[float]]
+        Dictionary mapping τ to list of displacements
+
+    Returns
+    -------
+    Tuple[Dict[int, float], Dict[int, float]]
+        Tuple of (mean_displacements, std_displacements) where each is a
+        dictionary mapping τ to the statistic
+    """
+    mean_displacement_per_tau = {
+        tau: np.mean(displacements)
+        for tau, displacements in displacement_per_tau.items()
+    }
+    std_displacement_per_tau = {
+        tau: np.std(displacements)
+        for tau, displacements in displacement_per_tau.items()
+    }
+    return mean_displacement_per_tau, std_displacement_per_tau
 
 
 def compute_dynamic_range(mean_displacement_per_tau):
```

Comment:
@edyoshikun , does this way of computing dynamic range make sense? You talked about the effect of outliers